# Pandas. Домашнее задание

В домашнем задании есть обязательные и дополнительные задачи.

*Обязательные задачи* нужно сделать, чтобы проверить, что вы действительно усвоили материал модуля. Они обязательны к загрузке на проверку преподавателю.

*Дополнительные задачи* для тех, кто чувствует, что нужно ещё немного потренироваться и порешать задачки по работе с Pandas. Вы также можете загрузить их на проверку преподавателю и получить по ним обратную связь.

Удачи!



---



---



**Цели домашнего задания:**
1. Научиться работать с описательными статистиками набора данных.
2. Научиться выбирать данные по условиям.
3. Научиться объединять несколько датафреймов в один.


**Что входит в домашнее задание:**
1. *Обязательные задачи*:

17 задач с использованием:
- набора рекламных данных, показывающих, щелкнул ли конкретный пользователь интернета рекламу на веб-сайте компании или нет,
- дополнительного набора данных, описывающего пользователей, посещающих веб-сайт компании.
2. *Дополнительные задачи*:

3 задачи для отработки исследования данных.


**Критерии оценки:**
- программа выдаёт верный ответ на заданном наборе данных;
- код читабелен: переменным даны осмысленные названия, отступы и правила расстановки пробелов соблюдены.

**Как отправить задание на проверку**

Выполните задачи, сохраните изменения при помощи опции Save and Checkpoint из вкладки меню File или кнопки Save and Checkpoint на панели инструментов. Отправьте через форму на платформе итоговый файл Jupyter Notebook (в формате .ipynb) или ссылку на него.



---


---



**Обязательные задачи**

В течение всего домашнего задания вы будете работать с несколькими наборами данных:
1. Набор рекламных данных, показывающих, щёлкнул ли конкретный пользователь рекламу на сайте компании или нет. Состоит из трёх разных файлов: `advertising_1.csv`, `advertising_2.csv`, `advertising_3.csv`.

Описание столбцов датасета:
- **Number**: Номер пользователя в системе.
- **Daily Time Spent on Site**: Среднее ежедневное время нахождения пользователя на сайте в минутах.
- **Daily Internet Usage**: Среднее ежедневное время нахождения пользователя в интернете в минутах.
- **Ad Topic Line**: Заголовок рекламного объявления.
- **Clicked on Ad**: Кликнул пользователь на рекламу или нет.

Дополнительный к рекламным данным датасет, который описывает пользователей, посещающих сайт компании. Состоит из одного файла: users.csv.

Описание столбцов датасета:
- **Number**: Номер пользователя в системе.
- **Age**: Возраст пользователя.
- **Male**: Является ли пользователь мужчиной.
- **City**: Город пользователя.
- **Country**: Страна пользователя.
- **Area Income**: Средний доход в месте жительства пользователя.


***Задача 1***

Загрузите таблицу из файла `advertising_1.csv` и сохраните её в датафрейм `adv1_df`. Укажите в качестве индекса столбец `Number`. Выведите на экран первые десять строк.


In [2]:
import pandas as pd

In [3]:
adv1_df = pd.read_csv('advertising_1.csv', index_col='Number')
print(adv1_df.head(10))

        Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
5                          68.95                256.09   
6                          80.23                193.77   
8                          74.15                245.89   
9                            NaN                225.58   
10                         59.99                226.74   
11                         88.91                208.36   
12                           NaN                131.76   
13                         74.53                   NaN   
14                         69.88                183.82   
15                         47.64                122.02   

                                Ad Topic Line  Clicked on Ad  
Number                                                        
5          Cloned 5thgeneration orchestration              0  
6          Monitored national standardization              0  
8       Triple-buffered reciprocal time-frame      

***Задача 2***

Выведите размер датафрейма `adv1_df`.

In [11]:
adv1_df.shape[0]

500

In [10]:
adv1_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 500 entries, 5 to 529
Data columns (total 4 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Daily Time Spent on Site  450 non-null    float64
 1   Daily Internet Usage      448 non-null    float64
 2   Ad Topic Line             500 non-null    object 
 3   Clicked on Ad             500 non-null    int64  
dtypes: float64(2), int64(1), object(1)
memory usage: 35.7+ KB


***Задача 3***

Выведите cреднее ежедневное время нахождения в интернете пользователя под номером 8.

In [5]:
print(adv1_df.loc[ 8, 'Daily Internet Usage'])

245.89


***Задача 4***

Загрузите таблицу из файла `advertising_2.csv` и сохраните её в датафрейм `adv2_df`. Укажите в качестве индекса стоблец `Number`. Выведите на экран строки для пользователей с номерами с 533-го по 536-й.

In [6]:
adv2_df = pd.read_csv('advertising_2.csv', index_col='Number')
print(adv2_df.loc[533:536])

        Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
533                        46.61                   NaN   
534                          NaN                162.03   
535                        65.10                118.10   
536                        53.44                   NaN   

                                Ad Topic Line  Clicked on Ad  
Number                                                        
533             Ergonomic methodical encoding              1  
534         Compatible dedicated productivity              0  
535            Up-sized real-time methodology              1  
536     Up-sized next generation architecture              1  


***Задача 5***

Выведите сводную таблицу описательных статистик для датафрейма `adv2_df`, вызвав необходимый метод на всём датафрейме.

Глядя на вывод, напишите, столбец с каким названием не содержит незаполненных значений.

*Подсказка:*  В сводной таблице не содержится информации о количестве строк всего датафрейма, поэтому для сначала распечатайте общее количество строк датафрейма.

*Рекомендация:* Кроме изученного метода `describe` для описания датафрейма и получения информации о пропущенных значениях полезен метод [`info`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html).

In [12]:
adv2_df.shape[0]

200

In [8]:
adv2_df.describe()

Daily Time Spent on Site  Daily Internet Usage  Clicked on Ad
count                184.000000            184.000000     200.000000
mean                  66.494728            186.809674       0.420000
std                   15.414725             43.844580       0.494797
min                   32.910000            108.030000       0.000000
25%                   52.805000            148.955000       0.000000
50%                   71.150000            194.170000       0.000000
75%                   78.352500            225.372500       1.000000
max                   91.430000            259.760000       1.000000

In [9]:
adv2_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 200 entries, 530 to 739
Data columns (total 4 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Daily Time Spent on Site  184 non-null    float64
 1   Daily Internet Usage      184 non-null    float64
 2   Ad Topic Line             200 non-null    object 
 3   Clicked on Ad             200 non-null    int64  
dtypes: float64(2), int64(1), object(1)
memory usage: 15.9+ KB


In [15]:
missing_values = adv2_df.isna().sum()
missing_values

Daily Time Spent on Site    16
Daily Internet Usage        16
Ad Topic Line                0
Clicked on Ad                0
dtype: int64

In [16]:
# Ваш код здесь
print('Не содержит незаполненных значений столбец:', ', '.join(missing_values[missing_values == 0].index.to_list()))

Не содержит незаполненных значений столбец: Ad Topic Line, Clicked on Ad


***Задача 6***

Добавьте к датафрейму `adv1_df` в конец датафрейм `adv2_df` с помощью метода `append`.

Сохраните объединенный датафрейм в новой переменной `adv12_df` и распечатайте его.

In [17]:
adv12_df = pd.concat([adv1_df, adv2_df])
adv12_df

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
5                          68.95                256.09   
6                          80.23                193.77   
8                          74.15                245.89   
9                            NaN                225.58   
10                         59.99                226.74   
...                          ...                   ...   
735                        65.53                190.17   
736                        61.87                   NaN   
737                        83.16                   NaN   
738                        44.11                121.24   
739                        56.57                131.98   

                                    Ad Topic Line  Clicked on Ad  
Number                                                            
5              Cloned 5thgeneration orchestration              0  
6              Monitored national standardization              0  
8           Triple-buffered reciprocal time-frame              0  
9                   Robust logistical utilization              0  
10                Sharable client-driven software              0  
...                                           ...            ...  
735        Pre-emptive client-server installation              0  
736     Multi-channeled attitude-oriented toolset              0  
737                 Decentralized 24hour approach              0  
738                Organic next generation matrix              1  
739          Multi-channeled non-volatile website              1  

[700 rows x 4 columns]

***Задача 7***

Для объединённого датафрейма `adv12_df` распечатайте среднее время, проводимое пользователем на сайте компании.

In [21]:
avg_time = adv12_df['Daily Time Spent on Site'].mean()
print(avg_time)

65.57268138801263


***Задача 8***

Из объединённого датафрейма `adv12_df` выберите и распечатайте строки только с теми пользователями, для которых неизвестно среднее ежедневное время, проводимое пользователем на сайте компании.

In [25]:
adv12_df[adv12_df['Daily Time Spent on Site'].isnull()]

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
9                            NaN                225.58   
12                           NaN                131.76   
19                           NaN                143.56   
27                           NaN                164.83   
34                           NaN                209.82   
...                          ...                   ...   
672                          NaN                235.97   
674                          NaN                126.44   
685                          NaN                200.71   
690                          NaN                226.69   
731                          NaN                225.05   

                                  Ad Topic Line  Clicked on Ad  
Number                                                          
9                 Robust logistical utilization              0  
12                     Reactive local challenge              1  
19                Grass-roots coherent extranet              1  
27        Mandatory disintermediate utilization              1  
34      Business-focused value-added definition              0  
...                                         ...            ...  
672            Customizable value-added project              0  
674                Reactive impactful challenge              1  
685               Focused multi-state workforce              0  
690               Total coherent superstructure              0  
731             Multi-tiered heuristic strategy              0  

[66 rows x 4 columns]

***Задача 9***

Загрузите таблицу из файла `advertising_3.csv` и сохраните её в датафрейм `adv3_df`. Укажите в качестве индекса стоблец `Number`. Выведите на экран столбцы `Ad Topic Line` и `Clicked on Ad`.

In [26]:
adv3_df = pd.read_csv("advertising_3.csv", index_col="Number")
adv3_df[["Ad Topic Line", "Clicked on Ad"]]

Ad Topic Line  Clicked on Ad
Number                                                             
740                 Distributed bifurcated challenge              0
741       Customizable zero-defect Internet solution              0
742     Self-enabling zero administration neural-net              1
744           Open-architected system-worthy ability              0
745               Quality-focused maximized extranet              1
...                                              ...            ...
999                    Front-line bifurcated ability              1
1000                   Fundamental modular algorithm              1
1001                 Grass-roots cohesive monitoring              1
1002                    Expanded intangible solution              1
1003            Proactive bandwidth-monitored policy              0

[251 rows x 2 columns]

***Задача 10***

Добавьте к датафрейму `adv12_df` в конец датафрейм `adv3_df` с помощью метода `concat`. Явно пропишите значение параметра `axis`.

Сохраните объединённый датафрейм в новой переменной `adv123_df` и распечатайте его.

In [27]:
adv123_df = pd.concat([adv12_df, adv3_df], axis=0)
adv123_df

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
5                          68.95                256.09   
6                          80.23                193.77   
8                          74.15                245.89   
9                            NaN                225.58   
10                         59.99                226.74   
...                          ...                   ...   
999                        43.70                173.01   
1000                       72.97                208.58   
1001                       51.30                   NaN   
1002                       51.63                120.37   
1003                       55.55                187.95   

                                Ad Topic Line  Clicked on Ad  
Number                                                        
5          Cloned 5thgeneration orchestration              0  
6          Monitored national standardization              0  
8       Triple-buffered reciprocal time-frame              0  
9               Robust logistical utilization              0  
10            Sharable client-driven software              0  
...                                       ...            ...  
999             Front-line bifurcated ability              1  
1000            Fundamental modular algorithm              1  
1001          Grass-roots cohesive monitoring              1  
1002             Expanded intangible solution              1  
1003     Proactive bandwidth-monitored policy              0  

[951 rows x 4 columns]

***Задача 11***

Для объединённого датафрейма `adv123_df` выберите только те строки, в которых пользователи кликнули на рекламное сообщение. Сохраните полученный результат в датафрейм `success_adv_df`, распечатайте размер нового датафрейма.

In [28]:
success_adv_df = adv123_df[adv123_df['Clicked on Ad'] == 1]
success_adv_df

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
12                           NaN                131.76   
15                         47.64                122.02   
17                         69.57                113.12   
19                           NaN                143.56   
20                         63.45                140.64   
...                          ...                   ...   
997                        69.17                123.62   
999                        43.70                173.01   
1000                       72.97                208.58   
1001                       51.30                   NaN   
1002                       51.63                120.37   

                                Ad Topic Line  Clicked on Ad  
Number                                                        
12                   Reactive local challenge              1  
15             Centralized neutral neural-net              1  
17      Centralized content-based focus group              1  
19              Grass-roots coherent extranet              1  
20         Persistent demand-driven interface              1  
...                                       ...            ...  
997             Extended leadingedge solution              1  
999             Front-line bifurcated ability              1  
1000            Fundamental modular algorithm              1  
1001          Grass-roots cohesive monitoring              1  
1002             Expanded intangible solution              1  

[475 rows x 4 columns]

***Задача 12***

Загрузите таблицу с данными пользователей из файла `users.csv` и сохраните её в датафрейм `users_df`. Индекс не указывайте.

Распечатайте сам датафрейм и сводную таблицу со статистиками для него.

In [29]:
users_df = pd.read_csv("users.csv")
users_df

Number   Age  Male            City                 Country  Area Income
0         5   NaN   0.0     Wrightburgh                 Tunisia     61833.90
1         6  31.0   NaN       West Jodi                   Nauru     68441.85
2         7  26.0   0.0        Davidton              San Marino     59785.94
3         8  29.0   1.0  West Terrifurt                   Italy     54806.18
4         9   NaN   0.0    South Manuel                 Iceland     73889.99
..      ...   ...   ...             ...                     ...          ...
975    1000  30.0   NaN       Duffystad                 Lebanon     71384.57
976    1001  45.0   1.0     New Darlene  Bosnia and Herzegovina     67782.17
977    1002  51.0   NaN   South Jessica                Mongolia     42415.72
978    1003  19.0   0.0     West Steven               Guatemala     41920.79
979    1004  26.0   0.0     Ronniemouth                  Brazil     29875.80

[980 rows x 6 columns]

In [30]:
users_df.describe()

Number         Age        Male   Area Income
count   980.000000  860.000000  868.000000    903.000000
mean    504.485714   36.073256    0.500000  55010.631938
std     287.956480    8.795992    0.500288  13423.710892
min       5.000000   19.000000    0.000000  13996.500000
25%     255.750000   29.000000    0.000000  46947.570000
50%     503.500000   35.000000    0.500000  57032.360000
75%     753.250000   42.000000    1.000000  65537.990000
max    1004.000000   61.000000    1.000000  79484.800000

***Задача 13***

Глядя на статистики `users_df`, укажите значения, удовлетворяющие следующим условиям:
1. Возраст самого молодого пользователя сайта: 19
2. Максимальный средний доход в месте жительства пользователя сайта: 79484.8
3. Средний возраст пользователя сайта: 36.073256
4. Количество заполненных значений для колонки, содержащей информацию о том, является пользователь мужчиной или нет: 868
5. Минимальный номер пользователя: 5

В этом задании не требуется выбирать по условиям и распечатывать значения, достаточно внимательно посмотреть на статистики и скопировать нужные значения согласно вопросам.

***Задача 14***

К полученному датафрейму с успешными рекламными сообщениями `success_adv_df` добавьте справа информацию из датафрейма с пользователями `users_df` и сохраните в датафрейм `success_full_df`. В полученном датафрейме должны остаться только те пользователи, номера которых есть в обоих датафреймах. Для этого воспользуйтесь методом `merge` и пропишите явно верные параметры.

Затем распечатайте датафрейм.

*Важно:*

Объединять датафреймы с помощью `merge` можно и по индексам, и по столбцам с разными названиями.
Так как общие значения содержатся в столбце `Number`, но в датафрейме `users_df` данный столбец является обычным столбцом, а в `success_adv_df` является индексом, перед тем как объединять, переведите столбец `Number` в `success_adv_df` в обычный столбец.

Для этого перед объединением пропишите:
```
success_adv_df = success_adv_df.reset_index()
```

[Подробнее о reset_index](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.reset_index.html)

In [31]:
success_adv_df = success_adv_df.reset_index()

In [32]:
success_full_df = pd.merge(success_adv_df, users_df, how='inner', on='Number')
success_full_df

Number  Daily Time Spent on Site  Daily Internet Usage  \
0        12                       NaN                131.76   
1        15                     47.64                122.02   
2        17                     69.57                113.12   
3        19                       NaN                143.56   
4        20                     63.45                140.64   
..      ...                       ...                   ...   
459     997                     69.17                123.62   
460     999                     43.70                173.01   
461    1000                     72.97                208.58   
462    1001                     51.30                   NaN   
463    1002                     51.63                120.37   

                             Ad Topic Line  Clicked on Ad   Age  Male  \
0                 Reactive local challenge              1  48.0   1.0   
1           Centralized neutral neural-net              1  49.0   NaN   
2    Centralized content-based focus group              1  48.0   1.0   
3            Grass-roots coherent extranet              1  33.0   0.0   
4       Persistent demand-driven interface              1  23.0   NaN   
..                                     ...            ...   ...   ...   
459          Extended leadingedge solution              1  40.0   NaN   
460          Front-line bifurcated ability              1  28.0   0.0   
461          Fundamental modular algorithm              1  30.0   NaN   
462        Grass-roots cohesive monitoring              1  45.0   1.0   
463           Expanded intangible solution              1  51.0   NaN   

                 City                 Country  Area Income  
0    Port Jefferybury               Australia     24593.33  
1     West Brandonton                   Qatar     45632.51  
2      West Katiefurt                   Egypt     51636.92  
3                 NaN                Barbados     30976.00  
4                 NaN                   Spain     52182.23  
..                ...                     ...          ...  
459       New Patrick              Montenegro     66193.81  
460      Nicholasland                 Mayotte     63126.96  
461         Duffystad                 Lebanon     71384.57  
462       New Darlene  Bosnia and Herzegovina     67782.17  
463     South Jessica                Mongolia     42415.72  

[464 rows x 10 columns]

***Задача 15***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, отобразите сводную таблицу со статистиками для строковых колонок (`Ad Topic Line`, `City`, `Country`).


In [34]:
success_full_df[["Ad Topic Line", "City", "Country"]].describe()

Ad Topic Line          City   Country
count                          464           420       418
unique                         464           412       203
top     Enhanced intangible portal  Williamsport  Ethiopia
freq                             1             2         7

In [33]:
success_full_df.describe(include='all')

Number  Daily Time Spent on Site  Daily Internet Usage  \
count    464.000000                416.000000            424.000000   
unique          NaN                       NaN                   NaN   
top             NaN                       NaN                   NaN   
freq            NaN                       NaN                   NaN   
mean     508.674569                 52.872716            145.614175   
std      294.724771                 12.881337             30.236861   
min       12.000000                 32.600000            104.780000   
25%      242.500000                 42.427500            123.592500   
50%      500.500000                 50.705000            138.860000   
75%      771.250000                 61.617500            161.160000   
max     1002.000000                 91.370000            269.960000   

                     Ad Topic Line  Clicked on Ad         Age        Male  \
count                          464          464.0  412.000000  411.000000   
unique                         464            NaN         NaN         NaN   
top     Enhanced intangible portal            NaN         NaN         NaN   
freq                             1            NaN         NaN         NaN   
mean                           NaN            1.0   40.252427    0.486618   
std                            NaN            0.0    8.860331    0.500430   
min                            NaN            1.0   19.000000    0.000000   
25%                            NaN            1.0   34.000000    0.000000   
50%                            NaN            1.0   40.000000    0.000000   
75%                            NaN            1.0   47.000000    1.000000   
max                            NaN            1.0   60.000000    1.000000   

                City   Country   Area Income  
count            420       418    425.000000  
unique           412       203           NaN  
top     Williamsport  Ethiopia           NaN  
freq               2         7           NaN  
mean             NaN       NaN  48272.526353  
std              NaN       NaN  13966.718364  
min              NaN       NaN  13996.500000  
25%              NaN       NaN  38745.290000  
50%              NaN       NaN  49111.470000  
75%              NaN       NaN  58183.040000  
max              NaN       NaN  78520.990000

***Задача 16***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, выведите страну и количество вхождений этой страны в датафрейм.


In [35]:
country_counts = success_full_df['Country'].value_counts()
country_counts

Country
Ethiopia                7
Turkey                  6
Australia               6
Liberia                 6
South Africa            5
                       ..
Haiti                   1
Colombia                1
United Arab Emirates    1
Malawi                  1
Comoros                 1
Name: count, Length: 203, dtype: int64

***Задача 17***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, выведите строки с данными пользователей из страны, название которой встречается чаще остальных.


In [37]:
success_full_df[success_full_df['Country'] == country_counts.index[0]]

Number  Daily Time Spent on Site  Daily Internet Usage  \
245     525                     46.61                136.18   
246     526                     63.26                120.46   
278     606                     52.13                118.27   
297     650                       NaN                137.97   
299     652                     46.43                137.20   
417     919                     41.86                128.62   
423     930                       NaN                   NaN   

                                       Ad Topic Line  Clicked on Ad   Age  \
245              Progressive intermediate throughput              1  42.0   
246                    Customizable holistic archive              1  29.0   
278                 Open-source even-keeled database              1  50.0   
297      Seamless bandwidth-monitored knowledge user              1  28.0   
299  Reverse-engineered background Graphic Interface              1  28.0   
417                      Open-source stable paradigm              1  39.0   
423           Front-line fresh-thinking installation              1  45.0   

     Male               City   Country  Area Income  
245   0.0      Catherinefort  Ethiopia     65856.74  
246   1.0         East Donna  Ethiopia     54787.37  
278   1.0  West Raymondmouth  Ethiopia     40926.93  
297   1.0          Markhaven  Ethiopia     39723.97  
299   1.0          Brianland  Ethiopia     53922.43  
417   1.0        Hawkinsbury  Ethiopia     53041.77  
423   0.0     Huffmanchester  Ethiopia     43241.88

***Задача 18***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, выведите строки с данными пользователей из страны, название которой встречается чаще остальных, они проводят в интернете в среднем больше 120 минут и их возраст меньше 30 лет.
Cохраните получившиеся строки в новый датафрейм `filtered_df`.


In [38]:
filtered_df = success_full_df[success_full_df['Country'] == country_counts.index[0]]
filtered_df = filtered_df[filtered_df['Daily Internet Usage'] > 120]
filtered_df = filtered_df[filtered_df['Age'] < 30]

filtered_df

Number  Daily Time Spent on Site  Daily Internet Usage  \
246     526                     63.26                120.46   
297     650                       NaN                137.97   
299     652                     46.43                137.20   

                                       Ad Topic Line  Clicked on Ad   Age  \
246                    Customizable holistic archive              1  29.0   
297      Seamless bandwidth-monitored knowledge user              1  28.0   
299  Reverse-engineered background Graphic Interface              1  28.0   

     Male        City   Country  Area Income  
246   1.0  East Donna  Ethiopia     54787.37  
297   1.0   Markhaven  Ethiopia     39723.97  
299   1.0   Brianland  Ethiopia     53922.43

***Задача 19***

Для получившегося датафрейма `filtered_df` выведите строки с данными пользователей, для которых либо не указано среднее время, проведённое на сайте, либо указанное время больше 55 минут. Cохраните получившиеся строки в новый датафрейм `filtered_df2`.


In [39]:
filtered_df2 = filtered_df[filtered_df['Daily Time Spent on Site'].isnull() | (filtered_df['Daily Time Spent on Site'] > 55)]
filtered_df2

Number  Daily Time Spent on Site  Daily Internet Usage  \
246     526                     63.26                120.46   
297     650                       NaN                137.97   

                                   Ad Topic Line  Clicked on Ad   Age  Male  \
246                Customizable holistic archive              1  29.0   1.0   
297  Seamless bandwidth-monitored knowledge user              1  28.0   1.0   

           City   Country  Area Income  
246  East Donna  Ethiopia     54787.37  
297   Markhaven  Ethiopia     39723.97

**Дополнительные задачи**

***Задача 20***

Для датафрейма `adv123_df` выведите строки с данными, заголовок рекламного сообщения в которых включает слово `model`.

Для этого вам понадобится метод [`contains`](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.contains.html).

Cохраните получившиеся строки в новый датафрейм `model_df`.


In [40]:
model_df = adv123_df[adv123_df['Ad Topic Line'].str.contains('model', case=False, na=False)]
model_df

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
62                         44.98                129.31   
119                        69.00                   NaN   
202                        78.57                   NaN   
211                        76.90                212.67   
314                          NaN                161.16   
350                        65.77                   NaN   
501                        81.10                216.49   
538                        91.43                209.91   
629                        78.17                192.27   
712                        81.90                225.47   
823                        81.98                212.88   
995                        35.79                165.62   

                              Ad Topic Line  Clicked on Ad  
Number                                                      
62                   Organized global model              1  
119                 Virtual composite model              0  
202           Exclusive client-driven model              0  
211           Synchronized multimedia model              0  
314              Extended interactive model              1  
350     Function-based fault-tolerant model              0  
501                Front-line dynamic model              0  
538                Organic motivating model              0  
629                  Automated mobile model              0  
712     Multi-channeled 3rdgeneration model              0  
823                Up-sized maximized model              0  
995          Enterprise-wide tangible model              1

***Задача 21***

Для получившегося датафрейма `model_df` выведите строки с данными пользователей, которые проводят на сайте больше среднего времени по всем пользователям в данном датафрейме. Cохраните получившиеся строки в новый датафрейм `model_popular_df`.


In [41]:
model_popular_df = model_df[model_df['Daily Time Spent on Site'] > model_df['Daily Time Spent on Site'].mean()]
model_popular_df

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
202                        78.57                   NaN   
211                        76.90                212.67   
501                        81.10                216.49   
538                        91.43                209.91   
629                        78.17                192.27   
712                        81.90                225.47   
823                        81.98                212.88   

                              Ad Topic Line  Clicked on Ad  
Number                                                      
202           Exclusive client-driven model              0  
211           Synchronized multimedia model              0  
501                Front-line dynamic model              0  
538                Organic motivating model              0  
629                  Automated mobile model              0  
712     Multi-channeled 3rdgeneration model              0  
823                Up-sized maximized model              0

***Задача 22***

Для объединённого датафрейма `model_popular_df` выведите строки с данными пользователей, номера которых равны 501, 712 и 823.

Для этого вам понадобится метод [`isin`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.isin.html).

*Подсказка:* Не забудьте, что номера пользователей находятся не в обычной колонке, а в колонке, являющейся индексом, поэтому вам необходимо получить значения индекса, а не обычной колонки.

In [42]:
model_popular_df[model_popular_df.index.isin([501, 712, 823])]

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
501                        81.10                216.49   
712                        81.90                225.47   
823                        81.98                212.88   

                              Ad Topic Line  Clicked on Ad  
Number                                                      
501                Front-line dynamic model              0  
712     Multi-channeled 3rdgeneration model              0  
823                Up-sized maximized model              0